iceoryx_hoofs/source/log/platform_building_blocks/console_logger.cpp

```diff
@@ -0,0 +1,232 @@
+// Copyright (c) 2019 by Robert Bosch GmbH. All rights reserved.
+// Copyright (c) 2021 - 2022 by Apex.AI Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include "iceoryx_hoofs/log/platform_building_blocks/console_logger.hpp"
+#include "iceoryx_hoofs/cxx/attributes.hpp"
+#include "iceoryx_hoofs/platform/time.hpp"
+
+#include <cstdio>
+#include <cstring>
+#include <ctime>
+
+namespace iox
+{
+namespace pbb
+{
+// NOLINTJUSTIFICATION see at declaration in header
+// NOLINTBEGIN(cppcoreguidelines-avoid-non-const-global-variables)
+std::atomic<LogLevel> ConsoleLogger::m_activeLogLevel{LogLevel::INFO};
+
+// NOLINTJUSTIFICATION see at declaration in header
+// NOLINTNEXTLINE(hicpp-avoid-c-arrays, cppcoreguidelines-avoid-c-arrays)
+thread_local char ConsoleLogger::m_buffer[ConsoleLogger::NULL_TERMINATED_BUFFER_SIZE]{0};
+
+thread_local uint32_t ConsoleLogger::m_bufferWriteIndex{0U};
+// NOLINTEND(cppcoreguidelines-avoid-non-const-global-variables)
+
+LogLevel ConsoleLogger::getLogLevel() noexcept
+{
+    return m_activeLogLevel.load(std::memory_order_relaxed);
+}
+
+// NOLINTJUSTIFICATION member access is required
+// NOLINTNEXTLINE(readability-convert-member-functions-to-static)
+void ConsoleLogger::setLogLevel(const LogLevel logLevel) noexcept
+{
+    m_activeLogLevel.store(logLevel, std::memory_order_relaxed);
+}
+
+void ConsoleLogger::createLogMessageHeader(const char* file,
+                                           const int line,
+                                           const char* function,
+                                           LogLevel logLevel) noexcept
+{
+    timespec timestamp{0, 0};
+    if (clock_gettime(CLOCK_REALTIME, &timestamp) != 0)
+    {
+        timestamp = {0, 0};
+        // intentionally do nothing since a timestamp from 01.01.1970 already indicates  an issue with the clock
```

Comment:
Yes, but this is more work and we could define that `01.01.1970` means invalid.

This will also fail if the system does not have a `CLOCK_REALTIME`.